sdk/core/Azure.Core/src/Pipeline/RetryPolicy.cs

```diff
@@ -55,216 +49,109 @@ public override ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpP
-                    }
-                }
-                catch (Exception ex)
-                {
-                    if (exceptions == null)
-                    {
-                        exceptions = new List<Exception>();
-                    }
-
-                    exceptions.Add(ex);
-
-                    lastException = ex;
-                }
+            AzureCorePipelineProcessor processor = new(pipeline);
 
+            try
+            {
                 if (async)
                 {
-                    await OnRequestSentAsync(message).ConfigureAwait(false);
+                    await _policy.ProcessAsync(message, processor, -1).ConfigureAwait(false);
                 }
                 else
                 {
-                    OnRequestSent(message);
-                }
-
-                var after = Stopwatch.GetTimestamp();
-                double elapsed = (after - before) / (double)Stopwatch.Frequency;
-
-                bool shouldRetry = false;
-
-                // We only invoke ShouldRetry for errors. If a user needs full control they can either override HttpPipelinePolicy directly
-                // or modify the ResponseClassifier.
-
-                if (lastException != null || (message.HasResponse && message.Response.IsError))
-                {
-                    shouldRetry = async ? await ShouldRetryAsync(message, lastException).ConfigureAwait(false) : ShouldRetry(message, lastException);
+                    _policy.Process(message, processor, -1);
                 }
-
-                if (shouldRetry)
-                {
-                    var retryAfter = message.HasResponse ? message.Response.Headers.RetryAfter : default;
-                    TimeSpan delay = async ? await GetNextDelayAsync(message, retryAfter).ConfigureAwait(false) : GetNextDelay(message, retryAfter);
-                    if (delay > TimeSpan.Zero)
-                    {
-                        if (async)
-                        {
-                            await WaitAsync(delay, message.CancellationToken).ConfigureAwait(false);
-                        }
-                        else
-                        {
-                            Wait(delay, message.CancellationToken);
-                        }
-                    }
-
-                    if (message.HasResponse)
-                    {
-                        // Dispose the content stream to free up a connection if the request has any
-                        message.Response.ContentStream?.Dispose();
-                    }
-
-                    message.RetryNumber++;
-                    AzureCoreEventSource.Singleton.RequestRetrying(message.Request.ClientRequestId, message.RetryNumber, elapsed);
-                    continue;
-                }
-
-                if (lastException != null)
+            }
+            catch (AggregateException e)
+            {
+                if (e.Message.StartsWith("Retry failed after"))
```

Comment:
This seems brittle, especially since the message doesn't seem to be defined in this class (or at least in this file).  Any way that we can set some other kind of data to flag a retry error, or call a method on the class that sets that message and let it decide?

Reply:
I added this because we have unit tests validating the text of the exception message with Azure.Core-specific logic.  You're totally right though -- let me see if there's another way. I think it would also be non-breaking to just update the exception message to make it helpful whether it originates in Azure.Core or ClientModel.

Reply:
Addressed in https://github.com/Azure/azure-sdk-for-net/pull/42329